sycl/plugins/unified_runtime/ur/adapters/hip/common.hpp

```diff
@@ -0,0 +1,176 @@
+// seem to be required and we're not using nvcc to build the UR HIP adapter so
+// add the translation function here
+#if defined(__HIP_PLATFORM_NVIDIA__) && !defined(__CUDACC__)
+inline static hipError_t
+hipArray3DCreate(hiparray *pHandle,
+                 const HIP_ARRAY3D_DESCRIPTOR *pAllocateArray) {
+  return hipCUResultTohipError(cuArray3DCreate(pHandle, pAllocateArray));
+}
+#endif
+
+// hipArray gets turned into cudaArray when using the HIP NVIDIA platform, and
+// some CUDA APIs use cudaArray* and others use CUarray, these two represent the
+// same type, however when building cudaArray appears as an opaque type, so it
+// needs to be explicitly casted to CUarray. In order for this to work for both
+// AMD and NVidia we introduce an second hipArray type that will be CUarray for
+// NVIDIA and hipArray* for AMD so that we can place the explicit casts when
+// necessary for NVIDIA and they will be no-ops for AMD.
+#if defined(__HIP_PLATFORM_NVIDIA__)
+typedef CUarray hipCUarray;
+#elif defined(__HIP_PLATFORM_AMD__)
+typedef hipArray *hipCUarray;
+#else
+#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
+#endif
+
+// Add missing HIP to CUDA defines
+#if defined(__HIP_PLATFORM_NVIDIA__)
+#define hipMemoryType CUmemorytype
+#define hipMemoryTypeHost CU_MEMORYTYPE_HOST
+#define hipMemoryTypeDevice CU_MEMORYTYPE_DEVICE
+#define hipMemoryTypeArray CU_MEMORYTYPE_ARRAY
+#define hipMemoryTypeUnified CU_MEMORYTYPE_UNIFIED
+#endif
+
+ur_result_t mapErrorUR(hipError_t Result);
+
+ur_result_t checkErrorUR(hipError_t Result, const char *Function, int Line,
+                         const char *File);
+
+#define UR_CHECK_ERROR(result)                                                 \
+  checkErrorUR(result, __func__, __LINE__, __FILE__)
+
+std::string getHipVersionString();
+
+constexpr size_t MaxMessageSize = 256;
+extern thread_local ur_result_t ErrorMessageCode;
+extern thread_local char ErrorMessage[MaxMessageSize];
+
+// Utility function for setting a message and warning
+[[maybe_unused]] void setErrorMessage(const char *Message,
+                                      ur_result_t ErrorCode);
+
+/// ------ Error handling, matching OpenCL plugin semantics.
+namespace detail {
+namespace ur {
+
+// Report error and no return (keeps compiler from printing warnings).
+// TODO: Probably change that to throw a catchable exception,
+//       but for now it is useful to see every failure.
+//
+[[noreturn]] void die(const char *pMessage);
+
+// Reports error messages
+void hipPrint(const char *pMessage);
+
+void assertion(bool Condition, const char *pMessage = nullptr);
+
+} // namespace ur
+} // namespace detail
+
+/// RAII object that calls the reference count release function on the held UR
+/// object on destruction.
+///
+/// The `dismiss` function stops the release from happening on destruction.
+template <typename T> class ReleaseGuard {
```

Comment:
It will already be enforced by the compiler as if we tried to use any type other than pointer type, the compilation will fail in this [lines](https://github.com/intel/llvm/blob/663042b04b63cb1830ae12b4442e12af4830764e/sycl/plugins/hip/pi_hip.cpp#L792) as they will give an invalid conversion, I think `enable_if` could give more clarity to this enforcement but I feel that it may also complicate this class than necessary.